src/globus_sdk/exc/api.py

```diff
@@ -0,0 +1,167 @@
+
+from .base import GlobusError
+from .err_info import ErrorInfoContainer
+
+log = logging.getLogger(__name__)
+
+
+class GlobusAPIError(GlobusError):
+    """
+    Wraps errors returned by a REST API.
+
+    :ivar http_status: HTTP status code (int)
+    :ivar code: Error code from the API (str),
+                or "Error" for unclassified errors
+    :ivar message: Error message from the API. In general, this will be more
+                   useful to developers, but there may be cases where it's
+                   suitable for display to end users.
+    """
+
+    MESSAGE_FIELDS = ["message", "detail"]
+    RECOGNIZED_AUTHZ_SCHEMES = ["bearer", "basic", "globus-goauthtoken"]
+
+    def __init__(self, r, *args, **kw):
+        self.http_status = r.status_code
+        # defaults, may be rewritten during parsing
+        self.code = "Error"
+        self.message = r.text
+
+        self._info: Optional[ErrorInfoContainer] = None
+        self._underlying_response = r
+        self._parse_response()
+        super().__init__(*self._get_args())
+
+    @property
+    def raw_json(self):
+        """
+        Get the verbatim error message received from a Globus API, interpreted
+        as JSON data
+
+        If the body cannot be loaded as JSON, this is None
+        """
+        r = self._underlying_response
+        if "Content-Type" in r.headers and (
+            "application/json" in r.headers["Content-Type"]
+        ):
+            try:
+                return r.json()
+            except ValueError:
+                log.error(
+                    "Error body could not be JSON decoded! "
+                    "This means the Content-Type is wrong, or the "
+                    "body is malformed!"
+                )
+                return None
+        else:
+            return None
+
+    @property
+    def raw_text(self):
+        """
+        Get the verbatim error message receved from a Globus API as a *string*
+        """
+        return self._underlying_response.text
+
+    @property
+    def info(self) -> ErrorInfoContainer:
+        """
+        An ``ErrorInfoContainer`` with parsed error data. The ``info`` of an error is
+        guaranteed to be present, but all of its contents may be falsey if the error
+        could not be parsed.
+        """
+        if self._info is None:
+            rawjson = self.raw_json
+            json_data = rawjson if isinstance(rawjson, dict) else None
+            self._info = ErrorInfoContainer(cast(Optional[Dict[str, Any]], json_data))
+        return self._info
+
+    def _get_request_authorization_scheme(self):
+        try:
+            authz_h = self._underlying_response.request.headers["Authorization"]
+            try:
+                authz_scheme = authz_h.split()[0]
+                if authz_scheme.lower() in self.RECOGNIZED_AUTHZ_SCHEMES:
+                    return authz_scheme
+            except IndexError:
+                pass
+        except KeyError:
+            pass
```

Comment:
This would be the same, right?
```suggestion
        try:
            authz_h = self._underlying_response.request.headers["Authorization"]
            authz_scheme = authz_h.split()[0]
            if authz_scheme.lower() in self.RECOGNIZED_AUTHZ_SCHEMES:
                return authz_scheme
        except (IndexError, KeyError):
            pass
```

Reply:
Yeah, I think the original was "stream of consciousness error handling". There's no need for separate handlers here.